v2/component/launcher.go

```diff
@@ -178,100 +182,125 @@ func NewLauncher(runtimeInfo string, options *LauncherOptions) (*Launcher, error
-		key = fmt.Sprintf(`{{$.inputs.artifacts['%s'].path}}`, k)
-		l.placeholderReplacements[key] = v.LocalArtifactFilePath
+		key = fmt.Sprintf(`{{$.inputs.artifacts['%s'].path}}`, name)
+		l.placeholderReplacements[key] = localPath
 
 		// Copy artifact to local storage.
+		copyErr := func(err error) error {
+			return fmt.Errorf("failed to download input artifact %q from remote storage URI %q: %w", name, inputArtifact.Uri, err)
+		}
 		// TODO: Selectively copy artifacts for which .path was actually specified
 		// on the command line.
-
-		blobKey, err := l.bucketConfig.keyFromURI(v.Artifact.GetUri())
+		blobKey, err := l.bucketConfig.keyFromURI(inputArtifact.Uri)
 		if err != nil {
-			return err
+			return copyErr(err)
 		}
-		r, err := bucket.NewReader(ctx, blobKey, nil)
-		if err != nil {
-			return err
-		}
-		defer r.Close()
 
-		if err := os.MkdirAll(path.Dir(v.LocalArtifactFilePath), 0644); err != nil {
-			return err
-		}
-		w, err := os.Create(v.LocalArtifactFilePath)
-		if err != nil {
-			return err
-		}
-		if _, err := io.Copy(w, r); err != nil {
-			return err
+		if err := downloadBlob(ctx, bucket, localPath, blobKey); err != nil {
+			return copyErr(err)
 		}
 	}
 
 	// Prepare input parameter placeholders.
-	for k, v := range l.runtimeInfo.InputParameters {
-		key := fmt.Sprintf(`{{$.inputs.parameters['%s']}}`, k)
-		l.placeholderReplacements[key] = v.ParameterValue
+	for name, parameter := range executorInput.Inputs.Parameters {
+		key := fmt.Sprintf(`{{$.inputs.parameters['%s']}}`, name)
+		switch t := parameter.Value.(type) {
+		case *pipeline_spec.Value_StringValue:
+			l.placeholderReplacements[key] = parameter.GetStringValue()
+		case *pipeline_spec.Value_DoubleValue:
+			l.placeholderReplacements[key] = strconv.FormatFloat(parameter.GetDoubleValue(), 'f', -1, 64)
+		case *pipeline_spec.Value_IntValue:
+			l.placeholderReplacements[key] = strconv.FormatInt(parameter.GetIntValue(), 10)
+		default:
+			return fmt.Errorf("unknown PipelineSpec Value type %T", t)
+		}
 	}
 
 	return nil
 }
 
-func (l *Launcher) prepareOutputs(ctx context.Context) error {
-	for k, v := range l.runtimeInfo.OutputParameters {
-		key := fmt.Sprintf(`{{$.outputs.parameters['%s'].output_file}}`, k)
-		l.placeholderReplacements[key] = v.FileOutputPath
-	}
+func (l *Launcher) prepareOutputs(ctx context.Context, executorInput *pipeline_spec.ExecutorInput) error {
+	for name, parameter := range executorInput.Outputs.Parameters {
+		key := fmt.Sprintf(`{{$.outputs.parameters['%s'].output_file}}`, name)
+		l.placeholderReplacements[key] = parameter.OutputFile
 
-	for k, v := range l.runtimeInfo.OutputArtifacts {
-		// TODO: sanitize k
-		v.LocalArtifactFilePath = path.Join("/tmp/kfp_launcher_outputs", k, "data")
+		dir := filepath.Dir(parameter.OutputFile)
+		if err := os.MkdirAll(dir, 0644); err != nil {
+			return fmt.Errorf("failed to create directory %q for output parameter %q: %w", dir, name, err)
```

Comment:
so we will never ask users to create the parent dir in their components? It's a great UX improvement!